crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3492,11 +3527,7 @@ impl<'db> TypeInferenceBuilder<'db> {
             ast::Expr::NumberLiteral(_literal) => Type::Todo,
             ast::Expr::BooleanLiteral(_literal) => Type::Todo,
 
-            // TODO: this may be a place we need to revisit with special forms.
-            ast::Expr::Subscript(subscript) => {
-                self.infer_subscript_expression(subscript);
-                Type::Todo
-            }
+            ast::Expr::Subscript(subscript) => self.infer_subscript_expression(subscript),
```

Comment:
Here is where we need the code to recognize special forms, but we should not be falling back to `infer_subscript_expression` here (that's for value expressions), instead we should have a dedicated `infer_subscript_type_expression` method, which should use `infer_type_expression` on the value and the index, and for now handle only the case where the value is `typing.Literal` special form, otherwise just return `Todo`.

Reply:
That makes sense. Just one question, should we use `infer_type_expression` for all indexes?
For example Literal in [here](https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-type_expression) is defined with an expression in the index. Others have `annotation_expression` in their grammar. So here I use `infer_type_expression` for other things but if it's Literal I use `infer_expression`.

My reasoning behind it was when the value is `True`. The `True` itself should not have any meaning when used alone in the type annotation.